flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/OneInputTransformationTranslator.java

```diff
@@ -41,33 +41,45 @@ public final class OneInputTransformationTranslator<IN, OUT>
     @Override
     public Collection<Integer> translateForBatchInternal(
             final OneInputTransformation<IN, OUT> transformation, final Context context) {
-        KeySelector<IN, ?> keySelector = transformation.getStateKeySelector();
         Collection<Integer> ids =
                 translateInternal(
                         transformation,
                         transformation.getOperatorFactory(),
                         transformation.getInputType(),
-                        keySelector,
+                        transformation.getStateKeySelector(),
                         transformation.getStateKeyType(),
                         context);
-        boolean isKeyed = keySelector != null;
-        if (isKeyed) {
-            BatchExecutionUtils.applyBatchExecutionSettings(
-                    transformation.getId(), context, StreamConfig.InputRequirement.SORTED);
-        }
+
+        maybeApplyBatchExecutionSettings(transformation, context);
 
         return ids;
     }
 
     @Override
     public Collection<Integer> translateForStreamingInternal(
             final OneInputTransformation<IN, OUT> transformation, final Context context) {
-        return translateInternal(
-                transformation,
-                transformation.getOperatorFactory(),
-                transformation.getInputType(),
-                transformation.getStateKeySelector(),
-                transformation.getStateKeyType(),
-                context);
+        Collection<Integer> ids =
+                translateInternal(
+                        transformation,
+                        transformation.getOperatorFactory(),
+                        transformation.getInputType(),
+                        transformation.getStateKeySelector(),
+                        transformation.getStateKeyType(),
+                        context);
+
+        if (transformation.isOutputOnlyAfterEndOfStream()) {
```

Comment:
There may be some operators defined on keyed stream that sort records internally. Maybe we should add another attribute to identify that case.